### PR TITLE
[RFC] document generic options

### PR DESCRIPTION
**Pull request by jreidinger**
https://trello.com/c/QNlLs9GW
not finished yet

### CODE REVIEW

File: doc/installation.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+```
+
+and network/config:
+```
+NETCONFIG_DNS_STATIC_SEARCH_LIST="suse.de"
+NETCONFIG_DNS_STATIC_SERVERS="192.168.122.1"
+```
+
+
+## inst_install_inf
```
Comment (mchf) [inline doc/installation.md]: in fact this is going to be removed (almost) completely soon.

File: doc/features.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+- allows to assign gateway for ipv4 and ipv6 and also assign via which device it can be reached
+- allows to enable ipv4 and ipv6 forwarding
+- allows manual edit of routing table
+
+
+Devices
+=======
+
+- allows to setup it without setup (for bondings), dhcp or static ( with ip, hostname, mask )
```
Comment (mchf) [inline doc/features.md]: and iBFT

File: doc/installation.md (hunk 1 above)
Comment (teclator) [inline doc/installation.md]: It should :), checking the interface I realized that most of the code is useless currently.

File: doc/features.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+IPv6
+==============
+
+- enable/disable it
+
+DHCP Client options
+====================
+
+- set hostname according to dhcp response
```
Comment (mchf) [inline doc/features.md]: i don't fully understand this note. If we talk about "Hostname to Send" then this is about hostname which is being send back to dhcp server.

File: doc/features.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
+Network Module Features
+-----------------------
+
+Network Services
+======================
+
+- wicked
+- Network Manager
+- no network service
```
Comment (mchf) [inline doc/features.md]: we're able to handle even old netconfig .. when it is installed. But i don't know if it is still doable in SLE-12.

File: doc/features.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+- set hostname according to dhcp response
+- set machine identifier for asking dhcp
+- allow to set specific option for hostname to send for dhcpd client ( how often it is used? )
+
+DNS
+===================
+
+- set hostname
+- set domain
```
Comment (mchf) [inline doc/features.md]: this is bad. hostname and domain is actually combined into FQDN and written into /etc/hostname. This feature is a bit controversial. See https://trello.com/c/so0amSNp/1176-sle-12-sp3-fate-yast2-network-clarify-hostname-setup

File: doc/features.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+- change hostname with dhcp ( looks same as in dhcp client options )
```
Comment (mchf) [inline doc/features.md]: no, that is something else

File: doc/features.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+- allows to assign gateway for ipv4 and ipv6 and also assign via which device it can be reached
+- allows to enable ipv4 and ipv6 forwarding
+- allows manual edit of routing table
+
+
+Devices
+=======
+
+- allows to setup it without setup (for bondings), dhcp or static ( with ip, hostname, mask and iBFT )
```
Comment (mchf) [inline doc/features.md]: ehm no. That note about iBFT belongs to "None" configuration - it means that configuration should be read from firmware of iBFT aware device

File: doc/features.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+- for infiniband it allows set IPoIB connected x datagram
+
+
+s390 Devices
+============
+
+- allows layer 2 and 3 devices
+- support channel ids for devices
+- allow layer 2 devices to be in bond
+- recognized types - qeth, lcs, ctc
```
Comment (mchf) [inline doc/features.md]: and iucv

File: doc/features.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+- for eth and token ring - allow to specify kernel module needed for device, if it is pcmci or usb and ethtool options
+- for each devices it can be specified when activate device, ifplugd priority, firewall zone and MTU
+- for bridge it allow to specify any physical devices ( so e.g. tap and tun are filtered out )
+- for bonding it allows to specify over which device it bonding
+- for wireless it allows to configure mode in which operating, encryption, essid of network and so on
+- for infiniband it allows set IPoIB connected x datagram
+
+
+s390 Devices
+============
```
Comment (mchf) [inline doc/features.md]: allows to set logical link address for L2 qeth device

File: doc/features.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+IPv6
+==============
+
+- enable/disable it
+
+DHCP Client options
+====================
+
+- send back to dhcp server hostname
```
Comment (mchf) [inline doc/features.md]: I'd rephrase it e.g. like this: "Set hostname which is send back to the DHCP server"

File: doc/features.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+- set hostname and domain, combining them into FQDN writting into /etc/hostname
+- change hostname with dhcp
+- assign hostname to local ip ( so own hostname is always resolvable )
+- modify /etc/resolv.conf, keep it manage automatic or modify it manually with selected values
+
+Routing
+=========
+
+- allows to assign gateway for ipv4 and ipv6 and also assign via which device it can be reached
```
Comment (mchf) [inline doc/features.md]: default gateway

File: doc/autoinstallation.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+`setup_before_proposal` has to be set to `true` as commented in the [autosetup section](#ausosetup), 
+more details about network 
+[here](https://www.suse.com/documentation/sles-12/singlehtml/book_autoyast/book_autoyast.html#CreateProfile.Network)
+
+#### ssh_settings_finish
+
+## Second Stage
+
+The _SecondStage_, also known as `continue` in the control file, is basically 
+the stage where the configuration should be definitly written.
```
Comment (jreidinger) [inline doc/autoinstallation.md]: I am missing here important option that keeps network config from first stage

File: doc/autoinstallation.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+### autoconfigure
+
+This client will read the 
+[desktop configuration files](https://yastgithubio.readthedocs.io/en/latest/autoyast-development/#desktop-configuration-file)
+and is very probably that will write the configuration of our network calling
+the lan_auto client with the information read in the networking section, but
+as commented, it depents on the information of the desktop files to do its job.
+
+
```
Comment (jreidinger) [inline doc/autoinstallation.md]: why so many empty lines?

File: doc/installation.md [hunk 14]
```diff
@@ -0,0 +1,98 @@
+## Introduction
+
+The workflow of the installation is customized via the [control
+file](https://github.com/yast/yast-installation/blob/master/doc/control-file.md), 
+and below we have the principal steps concerning network configuration:
+
+  - Linuxrc
+  - inst_install_inf (What is still needed)
+  - inst_dhcp
+  - manual_configuration (openSUSE only in Addons)
```
Comment (jreidinger) [inline doc/installation.md]: I think in SLE it is in registration, so in both you have chance to set it, not?

File: doc/autoinstallation.md (hunk 12 above)
Comment (teclator) [inline doc/autoinstallation.md]: It is described in network_finish method.

File: doc/autoinstallation.md (hunk 13 above)
Comment (teclator) [inline doc/autoinstallation.md]: Sorry, added the new line explicitly, fixing it.

File: doc/installation.md (hunk 14 above)
Comment (teclator) [inline doc/installation.md]: In SLE you can launch the `lan` client manually in `disks_activation`, `registration` and `addons' like you can see [here](https://gist.github.com/teclator/35a6aee2cd387605b6ef#final-implementation), so basically is the same without `registration`. I will rewrite it.

File: doc/autoinstallation.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+  - network_finish
+    - save_network
+
+#### network_finish
+
+This client calls `save_network` which copies udev rules and ifcfg files from
+from the running system which should be the linuxrc config and not the profile
+network configuration. 
+
+Take in account that the copy of the ifcfg files will be done only if
```
Comment (mchf) [inline doc/autoinstallation.md]: not only ifcfg files

File: doc/autoinstallation.md (hunk 13 above)
Comment (mchf) [inline doc/autoinstallation.md]: this client is not part of yast2-network but autoyast itself. Autoyast just generates a list of clients to call. The list is generated using "I-always-forget-option-name" from desktop file(s). So to get to the point, I think that this paragraph is useless and doesn't belong into network's documentation.

lan_auto client is THE client which uses AutoYaST api (Read, Write, Import, Export, ... and other calls) - entry point for AutoYaST either in autoinstallation even in running system.

File: doc/installation.md [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+  - finish 
+    - network_finish
+      - save_network
+
+Note: If you are interested in know more about installation process check 
+[this](https://github.com/yast/yast-installation/blob/master/doc/installation_overview.md) 
+documentation.
+
+### Linuxrc
```
Comment (mchf) [inline doc/installation.md]: I think that this is not intended to document linuxrc here.

File: doc/installation.md [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+[here](https://www.suse.com/documentation/sled-12/singlehtml/book_sle_deployment/book_sle_deployment.html#sec.i.yast2.network) 
+and also will be available in some steps during the installation (in Leap 42.2 
+only for Addons).
+
+## inst_finish
+
+In this step is when the configuration is really copied from the running
+system to the installed one. This client calls more specialized clients to
+procceed with the configuration, in case of networking it will call
+`network_finish` with will call `save_network`.
```
Comment (mchf) [inline doc/installation.md]: I'd mention differences in case of "upgrade"

File: doc/autoinstallation.md [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+like it is in the installation mode but the principal steps are not exactly
+the same, and also while in a normal installation the [Second
+Stage](https://www.suse.com/documentation/sles-12/singlehtml/book_autoyast/book_autoyast.html#overviewandconcept)
+is not needed, in autoinstallation the installation proccess is divided in
+two and in the _Second Stage_ is where and when the system configuration is
+really done.
+
+There are two ways to give a profile to _AutoYaST_, with (`autoyast` or with
+`autoyast2` parameters), the principal difference from networking point of 
+view is that `autoyast` will not fetch the profile which means that doesn't
```
Comment (mvidner) [inline doc/autoinstallation.md]: "`autoyast` will not fetch the profile" confusingly suggests that the profile is ignored. How about this:

> [the difference is that] `autoyast` leaves the fetching of the profile to YaST which means Linuxrc does not need to configure the network, while for `autoyast2` Linuxrc fetches the profile and may need need to configure the network.

File: doc/autoinstallation.md [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+#### network_finish
+
+This client calls `save_network` which copies udev rules and ifcfg files
+from the running system which should be the linuxrc config and not the profile
+network configuration. 
+
+Take in account that the copying of the ifcfg files will be done only if 
+`keep_install_network` has not been set to `false` and in case that we need to
+configure the network before _SecondStage_ then `setup_before_proposal` has to
+be set to `true` as commented in the [autosetup section](#ausosetup) more
```
Comment (mvidner) [inline doc/autoinstallation.md]: #au**t**osetup

File: doc/autoinstallation.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+details about network
+[here](https://www.suse.com/documentation/sles-12/singlehtml/book_autoyast/book_autoyast.html#CreateProfile.Network)
```
Comment (mvidner) [inline doc/autoinstallation.md]: This paragraph is hard to understand.

File: doc/autoinstallation.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+## Second Stage
+
+The _SecondStage_, also known as `continue` in the control file, is basically 
+the stage where the configuration should be definitively written.
+
+### autoconfigure
+
+This client will read the [desktop configuration
+files](https://yastgithubio.readthedocs.io/en/latest/autoyast-development/#desktop-configuration-file)
+and is very probably that will write the configuration of our network calling
```
Comment (mvidner) [inline doc/autoinstallation.md]: "Probably"? I would like my documentation to be more deterministic. Can we remove the uncertainty?

File: doc/autoinstallation.md [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+the `lan_auto` client with the information read in the `networking` section, but
+as mentioned, it depends on the information of the desktop files to do its
```
Comment (mvidner) [inline doc/autoinstallation.md]: "Depends on the desktop files." Well OK, but they change as rarely as the code. It's a different game than the AY profile or the control file. So let's assume them constant.

File: doc/installation.md [hunk 23]
```diff
@@ -0,0 +1,100 @@
+## Introduction
+
+The workflow of the installation is customized via the [control
+file](https://github.com/yast/yast-installation/blob/master/doc/control-file.md), 
+and below we have the principal steps concerning network configuration:
+
+  - Linuxrc
+  - inst_install_inf (What is still needed)
```
Comment (mvidner) [inline doc/installation.md]: "What is still needed" sounds like an editorial TODO. IMHO it can be left out of this overview

File: doc/installation.md [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+This client will try to configure dhcp in all the connected cards that haven't 
+been configured yet but only in the case that linuxrc has not activated one
+previously (i.e. with some parameter that implies a remote connection).
+
+## Manual Configuration (lan client)
+
+In case that the network was not configured then the **lan** client will be
+launched after the welcome dialog as you can see
+[here](https://www.suse.com/documentation/sled-12/singlehtml/book_sle_deployment/book_sle_deployment.html#sec.i.yast2.network) 
+and also will be available in some steps during the installation (in Leap 42.2
```
Comment (mvidner) [inline doc/installation.md]: "will be available in **some** steps" is not as clear and definite as I'd like. Does this depend on the control file? If yes, can we describe the typical choices?  Like: "for SLE, the registration needs the network, therefore it is set up at $this point, but for Leap..."

File: doc/autoinstallation.md (hunk 18 above)
Comment (teclator) [inline doc/autoinstallation.md]: Quite better now

File: doc/autoinstallation.md (hunk 22 above)
Comment (teclator) [inline doc/autoinstallation.md]: OK, make sense

**Comment by teclator** [comment]
I think that maybe make sense to split this PR in two, just keep this one with the generic features and a new one for installation and autoinstallation workflows. I could add more examples there and make it more clear if needed. 
Just an example of a typical bug: https://bugzilla.suse.com/show_bug.cgi?id=1010611
